bsdbpf/bsd_bpf_sniffer.go

```diff
@@ -0,0 +1,153 @@
+// +build darwin dragonfly freebsd netbsd openbsd
+
+/*
+ * This rough sketch of a golang API for a Berkeley packet filter sniffer for BSD variants.
+ * It provides only the capability to sniff ethernet frames... but could easily be extended for
+ * sending frames as well.
+ *
+ * This file was inspired by BSD licensed code from https://github.com/songgao/ether
+ *
+ * ( http://opensource.org/licenses/BSD-3-Clause )
+ *
+ * The bpf_wordalign function was borrowed... and readFrames was very much inspired by
+ * design I saw in songgao's ether git repository; the rest of the code I wrote myself ;-)
+ *
+ * Author: David Anthony Stainton
+ * License: BSD
+ *
+ */
+
+package bpf_sniffer
+
+import (
+	"fmt"
+	"github.com/google/gopacket"
+	"golang.org/x/sys/unix"
+	"syscall"
+	"time"
+	"unsafe"
+)
+
+const wordSize = int(unsafe.Sizeof(uintptr(0)))
+
+func bpf_wordalign(x int) int {
+	return (((x) + (wordSize - 1)) &^ (wordSize - 1))
+}
+
+type TimedFrame struct {
+	RawFrame  []byte
+	Timestamp time.Time
+}
+
+type BpfSniffer struct {
+	sniffDeveiceName string
+	bpfDeviceName string
+	fd       int
+	stopChan chan bool
+	readChan chan TimedFrame
+}
+
+func NewBpfSniffer(name string) *BpfSniffer {
+	return &BpfSniffer{
+		sniffDeviceName: name,
+		stopChan: make(chan bool, 0),
+		readChan: make(chan TimedFrame, 0),
+	}
+}
+
+func (b *BpfSniffer) pickBpfDevice() {
+	for i := 0; i < 99; i++ {
+		b.bpfDeviceName = fmt.Sprintf("/dev/bpf%d", i)
+		b.fd, err = syscall.Open(b.bpfDeviceName, syscall.O_RDWR, 0)
+		if err == nil {
+			break
+		}
+	}
+}
+
+// Init is used to initialize a BPF device for promiscuous sniffing.
+// It also starts a goroutine to continuously read frames.
+func (b *BpfSniffer) Init() error {
+	var err error
+	enable := 1
+
+	b.pickBpfDevice()
+
+	err = syscall.SetBpfInterface(b.fd, b.sniffDeviceName)
+	if err != nil {
+		return err
+	}
+
+	// turning Immediate mode off should make the snffer
+	// block when no packets are available.
+	err = syscall.SetBpfImmediate(b.fd, enable)
+	if err != nil {
+		return err
+	}
+
+	// preserves the link level source address...
+	// higher level protocol analyzers will not need this
+	err = syscall.SetBpfHeadercmpl(b.fd, enable)
+	if err != nil {
+		return err
+	}
+
+	// forces the interface into promiscuous mode
+	err = syscall.SetBpfPromisc(b.fd, enable)
+	if err != nil {
+		return err
+	}
+
+	go b.readFrames()
+	return nil
+}
+
+func (b *BpfSniffer) Stop() {
+	b.stopChan <- true
+}
+
+func (b *BpfSniffer) readFrames() {
```

Comment:
using a readFrames goroutine and readChan channel seems overly complex... instead, how about just calling lines 121-136 in ReadPacketData().

That way, if folks want to do things in a separate goroutine, they can, but if not, they don't have to.  It also makes error propagation from syscall.Read easier (it's just returned as an error from ReadPacketData), and there's no need  for stopChan.